### PR TITLE
BUG fixes range of pk variable pre-check

### LINKED ISSUE #9564: rv_sample accepts invalid input in scipy.stats
When passing values `(x_k, p_k)` to `stats.rv_discrete`, the only condition that is checked in `rv_sample` is whether `sum(p_k)` is one, but not whether the p_k are between zero and one

### Reproducing code example:

```
from scipy import stats
import numpy as np
custm = stats.rv_discrete(values=((1, 2, 3), (0.5, 1.2, -0.7))) # no error
```

### Scipy/Numpy/Python version information:
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.1.0 1.15.2 sys.version_info(major=3, minor=5, micro=4, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by akahard2dj**
closes #9564 

I applied pre-check for pk range.
previously pk is only checked by summation value (`np.sum(pk) == 1.0`).
In this feature, pk is checked by not only summation value but also each parameter range(0<pk<1).

Please review this PR.

### CODE REVIEW

File: scipy/stats/_distn_infrastructure.py [hunk 1]
```diff
@@ -3362,6 +3363,8 @@ def __init__(self, a=0, b=inf, name=None, badvalue=None,
 
         if len(xk) != len(pk):
             raise ValueError("xk and pk need to have the same length.")
+        if not np.allclose(np.size(np.where((np.array(pk) >= 0.0) & (np.array(pk) <= 1.0))), np.size(pk)):
```
Comment (ilayn) [inline scipy/stats/_distn_infrastructure.py]: This line is too long for PEP8, also you shouldn't compare integers with allclose. You can simply use `!=` and get rid of np.sizes
```python
if np.where((np.array(pk) >= 0.0) & (np.array(pk) <= 1.0))).size != len(pk):
    ...
```

**Review by ilayn — CHANGES_REQUESTED** [review]

File: scipy/stats/_distn_infrastructure.py [hunk 2]
```diff
@@ -2578,7 +2578,8 @@ class rv_discrete(rv_generic):
         The tolerance for the generic calculation of moments.
     values : tuple of two array_like, optional
         ``(xk, pk)`` where ``xk`` are integers with non-zero
-        probabilities ``pk``  with ``sum(pk) = 1``.
+        probabilities ``pk``  with ``sum(pk) = 1``. Note: pk should be
```
[the remark this answers is not in the document]
Comment (akahard2dj) [reply scipy/stats/_distn_infrastructure.py]: When I hear your opinion, a "Note" seems strange.
So I continue the sentence for description of pk conditions.

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (ilayn) [reply scipy/stats/_distn_infrastructure.py]: Ah, yes, it then reduces even further to

```
if all([x >= 0. for x in pk]):
    ....
```

Comment (akahard2dj) [reply scipy/stats/_distn_infrastructure.py]: Oh, that would be nice. 
I though about an algorithm too.

BUT this `if condition` is aim to raise a error for range of pk.
To simplify the code, I agree with all. 
In that case, if pk is negative, error is happen.
If pk is greater than 1.0, however, error is not happen (I know that `np.sum(pk)` can catch this error).

How do you think about?

Comment (ilayn) [reply scipy/stats/_distn_infrastructure.py]: Elements of `pk` cannot be negative if we test for `>= 0.` . So *first all elements are negative* condition is established. If elements are greater than `>= 1.` then their sum cannot be `<= 1.`

Comment (akahard2dj) [reply scipy/stats/_distn_infrastructure.py]: ok, I will change error message to

```python
if not all([x >= 0.0 for x in pk]):
    raise ValueError("pk cannot be negative values.")
```

Comment (chrisb83) [reply scipy/stats/_distn_infrastructure.py]: numpy.greater_equal might be an option as well

**Comment by charris** [comment]
Going to close, reopen to see if azure triggers.

**Comment by charris** [comment]
Yep, there is goes.

**Comment by akahard2dj** [comment]
All requests is reflected.

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (akahard2dj) [reply scipy/stats/_distn_infrastructure.py]: All requests are reflected.

**Comment by chrisb83** [comment]
test failures look unrelated to the change, see also issue #9595.

however, the current approach does not work for multi-dimensional input such as `pk = np.arange(15).reshape((3, 5))`.

Maybe `np.greater_equal(pk, 0).all()`?

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/_distn_infrastructure.py]: @akahard2dj should there be a unit test of the `ValueError` now (correctly) thrown?

Comment (akahard2dj) [reply scipy/stats/_distn_infrastructure.py]: Could you explain more detailed unit test of the `ValueError`?

**Comment by akahard2dj** [comment]
Yes! I did not notice that.
The input variable is an array-like, but I only check a python array.
I will fix that. Thanks @chrisb83

**Comment by chrisb83** [comment]
Thanks, addding a unit test would be great.

**Comment by akahard2dj** [comment]
Sure, No problem @chrisb83 .
I will follow a unit test including a multi-dimension `(xk, pk)` cases.

**Comment by akahard2dj** [comment]
@chrisb83 Test case is implemented in `test_distributions.py`.

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -1203,6 +1214,10 @@ def test_bad_input(self):
         pk = [1, 2, 3]
         assert_raises(ValueError, stats.rv_discrete, **dict(values=(xk, pk)))
 
+        xk = [1, 2, 3, 4, 5]
+        pk = [-0.3, -0.3, 0.3, 0.3, 0.4]
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: this test case does not distinguish whether the error is caused by negative values or sum != 1

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -1195,6 +1195,17 @@ def test_expect(self):
 
         assert_allclose(rv.expect(), np.sum(rv.xk * rv.pk), atol=1e-14)
 
+    def test_multidimension(self):
+        xk = np.array([[1,2,3,4],
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: to have a shorter test, use np.arange(12).reshape((3, 4)); (don't forget space after ,)

Comment (akahard2dj) [reply scipy/stats/tests/test_distributions.py]: Thanks. I will use `np.arange()` for this test case.

File: scipy/stats/tests/test_distributions.py (hunk 3 above)
Comment (akahard2dj) [reply scipy/stats/tests/test_distributions.py]: OK. I will do implementation for distinguishable test case (negative value and sum is 1).

**Comment by akahard2dj** [comment]
@charris requested items are fixed(shorter test, exact test case).
Please review this commit. :)

File: scipy/stats/_distn_infrastructure.py [hunk 5]
```diff
@@ -2577,8 +2577,8 @@ class rv_discrete(rv_generic):
     moment_tol : float, optional
         The tolerance for the generic calculation of moments.
     values : tuple of two array_like, optional
-        ``(xk, pk)`` where ``xk`` are integers with non-zero
-        probabilities ``pk``  with ``sum(pk) = 1``.
+        ``(xk, pk)`` where ``xk`` are integers with non-zero and
+        probabilities ``pk`` are ``sum(pk) = 1`` and bounded inside [0.0, 1.0].
```
Comment (chrisb83) [inline scipy/stats/_distn_infrastructure.py]: I would keep the old wording, the range should be clear

File: scipy/stats/_distn_infrastructure.py [hunk 6]
```diff
@@ -3362,6 +3362,8 @@ def __init__(self, a=0, b=inf, name=None, badvalue=None,
 
         if len(xk) != len(pk):
             raise ValueError("xk and pk need to have the same length.")
+        if np.less(pk, 0.0).any():
+            raise ValueError("pk cannot be negative values.")
```
Comment (chrisb83) [inline scipy/stats/_distn_infrastructure.py]: Rekord: All elements of pk must be non-negative.

File: scipy/stats/tests/test_distributions.py [hunk 7]
```diff
@@ -1195,6 +1195,15 @@ def test_expect(self):
 
         assert_allclose(rv.expect(), np.sum(rv.xk * rv.pk), atol=1e-14)
 
+    def test_multidimension(self):
+        xk = np.arange(12).reshape((3, 4))
+        pk = np.array([[0.1,0.1,0.15,0.05],
+                       [0.1,0.1,0.05,0.05],
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: PEP8 space after ,

**Review by chrisb83 — COMMENTED** [review]
Minor rewording + you need to rebase in master.

**Review by chrisb83 — APPROVED** [review]
Thanks, looks good now!

**Comment by chrisb83** [comment]
@ilayn hi, if your comments are addressed, feel free to merge.

**Review by ilayn — APPROVED** [review]

**Comment by ilayn** [comment]
Thanks @akahard2dj @chrisb83

**Pull request merged**